src/lib.rs

```diff
@@ -384,21 +384,50 @@ impl Drop for Context {
 
 pub struct Socket {
     sock: *mut c_void,
-    closed: bool
+    owned: bool,
 }
 
 unsafe impl Send for Socket {}
 
 impl Drop for Socket {
     fn drop(&mut self) {
-        match self.close_final() {
-            Ok(()) => { debug!("socket dropped") },
-            Err(e) => panic!(e)
+        if self.owned {
+            if unsafe { zmq_sys::zmq_close(self.sock) } == -1i32 {
+                panic!(errno_to_error());
+            } else {
+                debug!("socket dropped");
+            }
         }
     }
 }
 
 impl Socket {
+    /// Consume the Socket and return the raw socket pointer.
+    ///
+    /// Failure to close the raw socket manually or call `from_raw`
+    /// will lead to a memory leak.
+    pub fn into_raw(mut self) -> *mut c_void {
```

Comment:
I don't think `Deref` is appropriate here. I'd rather it be an explicit call to get the raw pointer. After a little looking around I think the appropriate type to follow would be [CString](https://doc.rust-lang.org/std/ffi/struct.CString.html), which has both a `into_raw()` and a `as_ptr()`. How's that sound?